File: practise/2.cpp
#include<iostream>
#include<algorithm>
#include<vector>

using namespace std;
int main()
{
	int t,n,T,m,i,a,maxdiff,mindiff,temp;
	vector<int> nice,hole;
	
	cin >> T;
		
	while(T--)
	{
		
		cin>>n>>m;
		
		for(i=0;i<n;i++)
		{
			cin >> a;
			nice.push_back(a);
		}
		
		for(i=0;i<m;i++)
		{

			cin >> a;
			hole.push_back(a);
		}
		
		sort(nice.begin(),nice.end());
		sort(hole.begin(),hole.end());
		
		maxdiff=1000000000;
		mindiff=0;
	
		while(maxdiff>mindiff)
		{
			temp=(maxdiff+mindiff)/2;
			t=0;
			for(i=0;i<n;)
			{
				if(abs(nice[i]-hole[t])>temp)
				{
					t++;
					int right=m-t;
					int rightn=n-i;
					if(rightn>right)break;
							
				}
				
				else
				{
					i++;					
					t++;
				}
			}
			
			if(i<n){
			mindiff=temp+1;
			}
			else
			{
				maxdiff=temp;
			}
			
		}
		
		cout << maxdiff << endl;
	
		nice.clear();
		hole.clear();	
	
	}
	
	
return 0;	

}

File: practise/2.java
/* Shubhang Verma */
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

#define ll long long int

bool checkavailable(ll t);
long long int a[200000];
long long int b[200000];
int n,m;



int main() {
    int t;
    cin>>t;
    while(t--)
    {
        cin>>n>>m;
        for(int i= 0 ; i< n ; i++)
            cin>>a[i];
        for(int i=0 ; i<m ; i++)
            cin>>b[i];
        sort(a,a+n);
        sort(b,b+m);
        ll low = 0;
        ll high = 1<<30;
        while(low < high)
        {
            ll mid = (low + high)/2;
            if(checkavailable(mid) == true)
                high = mid;
            else
                low = mid+1;
        }        
        cout<<low<<endl;
    }
    return 0;
}

bool checkavailable(ll t)
{
    int s = 0;
    int i = 0 ; 
    int j = 0;
    while(j < m && i < n)
    {
        if(abs(a[i] - b[j]) <= t)
        {
            i++;
            j++;
        }
        else
            j++;
    }
    if(i== n)
        return true;
    else
        return false;
}

